### PR TITLE
Create a preserve-whitespace option for the parser

### PR DESCRIPTION
**Pull request by sydneyli**
up until now, we've been ignoring all the whitespace information in parser.py-- let's create a flag so we can decide not to ignore it, and test that flag as much as we can.

~~Depends on #54 to pass some tests -- see `add whitespace parser tests` commit for this PR's diff.~~ EDIT: #54 merged!

### CODE REVIEW

**Comment by etingof** [comment]
Could you please rebase this patch?

Also:

```
./apacheconfig/lexer.py:172:80: E501 line too long (80 > 79 characters)
240
```

File: apacheconfig/parser.py [hunk 1]
```diff
@@ -150,12 +178,16 @@ def p_contents(self, p):
...
+                if self._preserve_whitespace:
+                    p[0] = p[1] + [p[2]]
+                else:
+                    p[0] = p[1]
             else:
                 # contents miditem
                 p[0] = p[1] + [p[2]]
         else:
-            if isinstance(p[1], str) and p[1].isspace():
+            if isinstance(p[1], str) and p[1].isspace()\
```
Comment (etingof) [inline apacheconfig/parser.py]: nit: may be reorder the operands to try cheaper ones first?

Comment (etingof) [inline apacheconfig/parser.py]: nit: may be parenthesis would look better than continuation (`\`)?

**Comment by sydneyli** [comment]
Rebased!

File: tests/unit/test_parser_whitespace.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+def _create_parser(options={}, start='contents'):
+    ApacheConfigLexer = make_lexer(**options)
+    ApacheConfigParser = make_parser(**options)
+    return ApacheConfigParser(ApacheConfigLexer(), start=start)
+
+class WhitespaceParserTestCase(unittest.TestCase):
+    def _test_cases(self, cases, tag='contents', options={}):
+        options['preserve_whitespace'] = True
+        parser = _create_parser(start=tag, options=options)
+        for x, expected in cases:
```
Comment (etingof) [inline tests/unit/test_parser_whitespace.py]: When we drop Py2 support, [subtests](https://docs.python.org/3/library/unittest.html#subtests) perhaps would be helpful here.

File: apacheconfig/parser.py [hunk 3]
```diff
@@ -53,6 +59,8 @@ def p_includeoptional(self, p):
         """includeoptional : APACHEINCLUDEOPTIONAL
         """
         # lexer returns [includeoptional, whitespace, value]
+        if self._preserve_whitespace:
+            p[0] = ['includeoptional'] + p[1]
         p[0] = ['includeoptional', p[1][2]]
```
Comment (etingof) [inline apacheconfig/parser.py]: Will this assignment override the previous one at line 63?

File: tests/unit/test_parser_whitespace.py (hunk 2 above)
Comment (sydneyli) [reply tests/unit/test_parser_whitespace.py]: yeah! I created an issue to track this: #57

File: apacheconfig/parser.py (hunk 3 above)
Comment (sydneyli) [reply apacheconfig/parser.py]: Oh, great catch! This got messed up when I split up the larger PR. Added a test as well!

**Pull request merged**